fastwarc/fastwarc/cli.py

```diff
@@ -167,13 +169,87 @@ def check(infile, decompress_alg, verify_payloads, quiet, output):
+@click.option('--headers', is_flag=True, help='output only record (and HTTP) headers')
+def extract(infile, offset, payload, headers):
+    """Extract WARC record by offset.
+    """
+    compl_record = not (payload or headers)
+    with open(infile, 'rb') as stream:
+        stream.seek(offset)
+        record = next(iter(ArchiveIterator(stream)))
+        if headers or compl_record:
+            ostream = PythonIOStreamAdapter(sys.stdout.buffer)
+            record.headers.write(ostream)
+            sys.stdout.buffer.write(b'\r\n')
+            if record.is_http:
+                record.http_headers.write(ostream)
+                sys.stdout.buffer.write(b'\r\n')
+        if payload or compl_record:
+            buf = record.reader.read(4096)
+            while buf:
+                sys.stdout.buffer.write(buf)
+                buf = record.reader.read(4096)
+
 
 @main.group()
 def benchmark():
     """Benchmark FastWARC performance."""
     return 0
 
 
+def _index_record(output, fields, record, next_record_offset, file_name):
+    idx = OrderedDict()
+    for f in fields:
+        v = None
+        if f == 'offset':
+            v = str(record.stream_pos)
+        elif f == 'length':
+            v = str(next_record_offset - record.stream_pos)
+        elif f == 'filename':
+            v = file_name
+        elif f == 'http:status':
+            if (record.record_type & (WarcRecordType.response | WarcRecordType.revisit)) and record.http_headers:
+                idx[f] = str(record.http_headers.status_code)
+        elif f.startswith('http:'):
+            if record.http_headers:
+                v = record.http_headers.get(f[5:])
+        elif f == 'rec.content_length':
+            v = record.content_length
+        else:
+            v = record.headers.get(f)
+        if v is not None:
+            idx[f] = v
+    output.write(json.dumps(idx) + '\n')
+
+
+@main.command()
+@click.argument('infiles', type=click.Path(dir_okay=False, exists=True), nargs=-1)
+@click.option('-o', '--output', type=click.File('w'), default=sys.stdout,
+              help='Output file, default is stdout')
+@click.option('-f', '--fields', type=click.STRING,
+              default='offset,warc-type,warc-target-uri',
+              help='comma-separated list of indexed fields, eg. "offset", "length", "filename", '
+                   '"http:status", "http:<http-header>", or "<warc-record-header>"')
+def index(infiles, output, fields):
+    """Index WARC records into CDXJ.
+    """
+    fields = fields.split(',')
+    for infile in infiles:
+        with open(infile, 'rb') as stream:
+            last_record = None
+            for record in ArchiveIterator(stream):
+                if last_record is not None:
+                    _index_record(output, fields, last_record, record.stream_pos, infile)
+                last_record = record
+            if last_record:
+                _index_record(output, fields, last_record, stream.tell(), infile)
```

Comment:
This is confusing. I would prefer a `length` property on `WarcRecord` that returns the entire record length.

Reply:
Definitely agree. But not today. Feel free to implement this, I'd have to look how to do this anyway. Can be tricky because the entire remaining record needs to be consumed to get the length in the raw WARC file including gzip trailers and alike.

Reply:
Do you really want the length of the input stream segment? Wouldn't the length of the extracted record itself make more sense?

Reply:
No. The "length" must be the length of the WARC record in the WARC file. Offset and length are used to fetch single WARC records:
```
curl -s -r931-$((931+12680-1)) http://commoncrawl.s3.amazonaws.com/crawl-data/CC-NEWS/2021/09/CC-NEWS-20210930113548-00741.warc.gz | gzip -dc
```

Reply:
Okay, that's not possible without read-ahead. In that case, I'd suggest:

```python
from fastwarc.stream_io import _buf_reader_py_test_reset_limit, _buf_reader_py_test_set_limit

# ...

with open(infile, 'rb') as stream:
    for record in ArchiveIterator(stream):
        record.reader.consume()
        _buf_reader_py_test_reset_limit(record.reader)
        _index_record(output, fields, preserve_multi_header, record, record.reader.tell(), infile)
        _buf_reader_py_test_set_limit(record.reader, 0)
```

Reply:
Done. Yes, it works. But only in the sense that it gives the same offsets than the solution using the last record. Comparing the offsets of `TOOL index -fwarc-type,warc-target-uri,offset,length CC-NEWS-20210930113548-00741.warc.gz` for warcio and fastwarc, points to an issue where sometimes the value of record.stream_pos is not increased. Could you have a look?

```
12381,12382c12381,12382
< {"warc-type": "response", "warc-target-uri": "https://www.themarketsdaily.com/2021/09/30/ishares-sp-500-etf-nysearcaivv-sees-strong-trading-volume.html", "offset": "232757027", "length": "12286"}
< {"warc-type": "request", "warc-target-uri": "https://www.agenzianova.com/campania/6155a1e2e36aa9.17205456/3620963/2021-09-29/napoli-approvati-in-giunta-due-interventi-sulle-scuole", "offset": "232769313", "length": "514"}
---
> {"warc-type": "response", "warc-target-uri": "https://www.themarketsdaily.com/2021/09/30/ishares-sp-500-etf-nysearcaivv-sees-strong-trading-volume.html", "offset": "232757027", "length": "0"}
> {"warc-type": "request", "warc-target-uri": "https://www.agenzianova.com/campania/6155a1e2e36aa9.17205456/3620963/2021-09-29/napoli-approvati-in-giunta-due-interventi-sulle-scuole", "offset": "232757027", "length": "12800"}
```

Reply:
Hmm. That probably happens when a record is larger than a single GZip member. That needs some more investigating, but isn't something that can be fixed in this PR. FastWARC reads ahead in the stream in fixed block sizes or until the end of a member is found. The BufferedReader on top of that doesn't know anything about GZip members, so there is quite some magic going on to calculate raw stream offsets from the reader offsets. Something must go wrong with in the offset calculation:

https://github.com/chatnoir-eu/chatnoir-resiliparse/blob/develop/fastwarc/fastwarc/stream_io.pyx#L435

https://github.com/chatnoir-eu/chatnoir-resiliparse/blob/develop/fastwarc/fastwarc/stream_io.pyx#L924

Reply:
Found the reason:

55800cd382c5ef9ccae5e97526a3b700db7ef1fd

Please check if everything works now.

The problem was that I didn't update the stream position if the internal buffer emptied between records while consuming newlines. I only did that for uncompressed streams. Unfortunately, I cannot build test cases for this, since it's a very low-chance event that depends on the sizes of all previous records. Simply extracting the affected records into a small test WARC therefore won't work and I obviously cannot add the full WARC to the repo. I already have tests that check that each record has a higher offset than the previous one, but they didn't trigger on the test WARCs. If you happen to have a WARC where this occurs within the first few kB, please add a small sample of it to `tests/data` and submit a PR.

Reply:
NVM, I have tests now. Imanaged to create two synthetic WARCs that reproduce the issue by containing records that are exactly the internal buffer size: 4ef411983d647e33c3f9372ac0908c8ea990d45b